File: Ballast.Core/src/Client/models/offset-coordinates.ts
import { IAxialCoordinates, AxialCoordinates } from './axial-coordinates';
import { ICubicCoordinates, CubicCoordinates } from './cubic-coordinates';

export interface IOffsetCoordinates  {
    col: number;
    row: number;
}

export class OffsetCoordinates implements IOffsetCoordinates {

    public readonly col: number;
    public readonly row: number;

    private constructor(state: IOffsetCoordinates) {
        this.col = state.col;
        this.row = state.row;        
        if (this.col == -0)
            this.col = 0;
        if (this.row == -0)
            this.row = 0;
    }

    public static fromObject(object: IOffsetCoordinates) {
        return new OffsetCoordinates(object);
    }

    public static fromAxial(object: IAxialCoordinates) {
        // Bitwise AND (& 1) to get 0 for even or 1 for odd column offset
        let col = object.x + (object.z - (object.z & 1)) / 2;
        let row = object.z;
        return new OffsetCoordinates({col: col, row: row});
    }

    public static fromCubic(object: ICubicCoordinates) {
        return OffsetCoordinates.fromAxial(object);
    }

    public static fromOrderedPair(orderedPair: number[]) {
        if (orderedPair.length < 2)
            throw new Error('Length of ordered pair must be 2 (or greater)');
            let col = orderedPair[0]; // x
            let row = orderedPair[1]; // z
        return OffsetCoordinates.fromObject({ col: col, row: row });
    }

    public toAxial() {
        return AxialCoordinates.fromOffset(this);
    }

    public toCubic() {
        return CubicCoordinates.fromOffset(this);
    }

    public toOrderedPair() {
        return [this.col, this.row];
    }

}

File: ballast-core/dotnet/src/domain/models/OffsetCoordinates.cs
using System;

namespace Ballast.Core.Domain.Models
{
    public class OffsetCoordinates
    {

        public int Row { get; private set; }
        public int Col { get; private set; }

        public OffsetCoordinates(int col, int row) 
        {
            Col = col;
            Row = row;        
        }

        public static OffsetCoordinates FromAxial(AxialCoordinates axial) 
        {
            // Bitwise AND (& 1) to get 0 for even or 1 for odd column offset
            var col = axial.X + (axial.Z - (axial.Z & 1)) / 2;
            var row = axial.Z;
            return new OffsetCoordinates(col: col, row: row);
        }
        
        public static OffsetCoordinates FromCubic(CubicCoordinates cubic) =>
            OffsetCoordinates.FromAxial(
                AxialCoordinates.FromCubic(cubic)
            );
        
        public static OffsetCoordinates FromOrderedPair(int[] orderedPair) {
            if (orderedPair.Length < 2)
                throw new Exception("Length of ordered pair must be 2 (or greater)");
                var col = orderedPair[0]; // x
                var row = orderedPair[1]; // z
            return new OffsetCoordinates(col: col, row: row);
        }

        public AxialCoordinates ToAxial() => AxialCoordinates.FromOffset(this);

        public CubicCoordinates ToCubic() => CubicCoordinates.FromOffset(this);

        public int[] ToOrderedPair() => new int[] { Col, Row };

    }
}